src/core/mantle/evaluator/simple/expression/expression.ts

```diff
@@ -24,11 +25,23 @@ export class Expression {
     this._expression = expression;
     this._feature = feature;
     let literalJP: JPLiteral[] = [];
-    expression = replaceDefines(expression, defines);
-    [expression, literalJP] = replaceVariables(
-      removeBackslashes(expression),
-      this._feature?.properties,
-    );
+
+    const hasJSONPath = /\$[a-zA-Z[.]/.test(expression);
+
+    const cachedReplacedVariables = REPLACED_VARIABLES_CACHE.get(expression);
+    // JSONPath returns simple literal so we can not cache in here
+    if (cachedReplacedVariables && !hasJSONPath) {
+      [expression, literalJP] = cachedReplacedVariables;
+    } else {
+      const originalExpression = expression;
+      expression = replaceDefines(expression, defines);
+      [expression, literalJP] = replaceVariables(
+        removeBackslashes(expression),
+        this._feature?.properties,
+        hasJSONPath,
+      );
+      REPLACED_VARIABLES_CACHE.set(originalExpression, [expression, literalJP]);
+    }
```

Comment:
I think we can cache replacedVariable without JSONPath. JSONPath is evaluated in replacedVariables, also it returns evaluated result, so we can not cache result value.